BUG: allow MaskedArray min/max with dtype=object

benburrill opened the pull request:
Fixes #9103.  There are actually 2 bugs here:
- The `AttributeError` is fixed by checking if the result of min/max is an `ndarray` before calling `.view()`.
- The `TypeError` (which is what you actually get when you run the [#9103](https://github.com/numpy/numpy/issues/9103#issue-228130721) reproducer, even though it is reported as the AttributeError) is fixed by adding a universal infinity filler object which (preferentially) compares greater than all other objects other than itself.  Previously, the string `'?'` was used as the default filler for object min/max, which was wrong even for the case of an array of string objects.

In the process of fixing this bug, I also found several other related issues by searching for uses of `.view()`.  I was initially planning to fix those in this PR, but I think some of them may require a bit more discussion.  I plan to open a new issue about that later.

<!--         ----------------------------------------------------------------
                MAKE SURE YOUR PR GETS THE ATTENTION IT DESERVES!
                ----------------------------------------------------------------

*  FORMAT IT RIGHT:
      https://www.numpy.org/devdocs/dev/development_workflow.html#writing-the-commit-message

*  IF IT'S A NEW FEATURE OR API CHANGE, TEST THE WATERS:
      https://www.numpy.org/devdocs/dev/development_workflow.html#get-the-mailing-list-s-opinion

*  HIT ALL THE GUIDELINES:
      https://numpy.org/devdocs/dev/index.html#guidelines

*  WHAT TO DO IF WE HAVEN'T GOTTEN BACK TO YOU:
      https://www.numpy.org/devdocs/dev/development_workflow.html#getting-your-pr-reviewed
-->

benburrill committed:
BUG: allow MaskedArray min/max with dtype=object

Fixes #9103

benburrill committed:
TST: improve tests of min/max for MAs of objects

Tests the case of arrays of 0-D arrays, and check for identity rather
than equality.

benburrill commented on numpy/ma/core.py:
```diff
@@ -5942,8 +5942,9 @@ def min(self, axis=None, out=None, fill_value=None, keepdims=np._NoValue):
             result = self.filled(fill_value).min(
-                axis=axis, out=out, **kwargs).view(type(self))
-            if result.ndim:
+                axis=axis, out=out, **kwargs)
```
Hmm... even with the contemporary numpy 1.13.0, I'm seeing the `TypeError` rather than the `AttributeError` on the #9103 reproducer.  I only get the `AttributeError` if I remove the mask or set a fill_value.  I'd test older versions, but I'm running into some segfaults.

benburrill replied:
I've been trying to think if there is any "good" way to fix the `TypeError` issue, besides the obvious way of creating a universal infinity object that compares greater to any object regardless of type, and putting that in `min_filler`, but really I don't think there is (short of completely reimplementing min).  That said, it should be a very easy fix to create such an object, so I'd be willing to add that to the PR if you think it is an acceptable solution.

benburrill replied:
I went ahead with my proposal to add a universal infinity filler to fix the TypeError issue.

benburrill committed:
TST: parametrize test_minmax_object_dtype

tylerjereddy commented:
I haven't thought too much more about this, but if the discussion above is suggesting that this PR doesn't fix the reproducer in the original ticket, it might be sensible to at least not have this PR set to automatically close/address the ticket, etc.

Then the scope might be clearer/a bit less confusing--it may very well be just fine to address the separate issue here without closing that ticket.

benburrill commented:
That's reasonable.  I changed it to no longer close #9103.

benburrill commented on numpy/ma/core.py:
```diff
@@ -5942,8 +5942,9 @@ def min(self, axis=None, out=None, fill_value=None, keepdims=np._NoValue):
-                axis=axis, out=out, **kwargs).view(type(self))
-            if result.ndim:
+                axis=axis, out=out, **kwargs)
+            if isinstance(result, ndarray) and result.ndim:
```
There is an (extreme) edge case I am aware of that my code fails to address: ideally, I think that

```python
x = np.array([1, 2])
assert np.fromiter([x], dtype=object).view(np.ma.MaskedArray).min() is x
```

should pass, but with my change, x will get erroneously converted to a masked array when returned by `min`.

This is identical to the old behavior, so it's not necessarily a problem, but it is a failure of my assumptions: initially when working on this PR, I thought I only needed to worry about 0-D arrays, since arrays of higher dimension are not orderable (and so you'd just get a ValueError when calling `min()`, as you should).  But of course I forgot to consider that `min()` will succeed regardless if there is only one item.

I think this *could* be fixed by checking for the special case of arrays containing only one item, but I'm not sure if this is actually worth fixing.  Thoughts?

benburrill committed:
BUG: universal infinity min/max filler for objects

Adds _UniversalInfinity object which (preferentially) compares greater
than all other objects other than itself.  This is used as the default
min_filler.  Previously, min filled masked values with the string '?',
which caused a TypeError when the array contained objects that could not
be compared with str.
Likewise, negative universal infinity is used as the max_filler.
See #9103.

benburrill committed:
TST: test #9103 reproducer specifically

NOTE: The reproducer given in #9103, although reported as producing an
AttributeError, actually produced a TypeError.  The AttributeError was a
secondary issue which occurs when there is no mask.  Because there were
2 bugs associated with the issue (one of was unnoticed in my previous
attempt to fix the issue), we test the reproducer specifically just as
confirmation that the issue is fully resolved and I'm not missing
anything else.
This test *should* be redundant with the added test_minmax_object_dtype
tests, but it's not exactly the same.

benburrill commented:
I have added a proposed fix to the `TypeError` issue, so I have changed it back to close #9103.

benburrill commented:
I would appreciate additional feedback, especially regarding the introduction of the `_UniversalInfinity` filler object.  Let me know if you want to see more tests, etc.
